Integration of more descriptors

JonasSchaub opened the pull request.

ManuelSchauer committed:
Added three Descriptors (ATOM_COUNT, H_BOND_ACCEPTOR_COUNT, H_BOND_DONOR_COUNT) along with tests.

ManuelSchauer committed:
Added descriptors along with tests, better comments.

ManuelSchauer committed:
Added descriptors along with tests, changed Aromaticity Model in Test classes.

ManuelSchauer committed:
Added new descriptors along with tests.

ManuelSchauer committed:
Added new descriptors along with tests.

ManuelSchauer committed:
Added new descriptors along with tests.

JonasSchaub commented:
@ManuelSchauer just a note, more descriptive commit messages would be preferable (not always the same one), e.g. name the specific descriptors added in a specific commit.

ManuelSchauer committed:
Added method createMoleculeWithExplicitHydrogens along with a test, implementation of the method in A_LOG_P und X_LOG_P descriptor. Changed tests to fit the new implementation.

ManuelSchauer committed:
Added first implementation of method setAromaticity along with a test. First implementation in test_AROMATIC_ATOMS_COUNT.

ManuelSchauer committed:
Added descriptors along with tests(AUTOCORRELATION_CHARGE, AUTOCORRELATION_MASS, AUTOCORRELATION_POLARIZABILITY, FRAGMENT_COMPLEXITY, CHI_CHAIN, CHI_CLUSTER, CHI_PATH_CLUSTER, CHI_PATH). Changed Aromaticity Models to match new API.

ManuelSchauer committed:
Added three new descriptors along with tests(FRACTIONAL_PSA, LARGEST_PI_SYSTEM, SMALL_RING).

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -54,44 +101,533 @@ public enum Descriptor {
+     * Excluding terminal bonds.
+     * TODO: Include terminal bonds? Exclude Amide C-N bonds?
+     */
+    ROTATABLE_BONDS_COUNT,
```
If I understand the documentation on the default parameters correctly, amide bonds *are* counted by default (exclude: false).
https://cdk.github.io/cdk/latest/docs/api/org/openscience/cdk/qsar/descriptors/molecular/RotatableBondsCountDescriptor.html

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -54,44 +101,533 @@ public enum Descriptor {
+     * C3SP3 - singly bound carbon bound to three other carbons
+     * C4SP3 - singly bound carbon bound to four other carbons
+     */
+    CARBON_TYPES,
```
as discussed today, line breaks in JavaDoc are ignored when it is parsed. You need to use \<br\> to start a new line.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -54,44 +101,533 @@ public enum Descriptor {
+     * 4. nAromBlocks - total number of aromatically connected components
+     * TODO: Expand to specific ring sizes?
+     */
+    SMALL_RING;
```
You could use editor folds to group descriptors of a similar category.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -54,44 +101,533 @@ public enum Descriptor {
-        descriptorToComponentNumberMap.put(MOLECULER_WEIGHT, 1);
-        descriptorToCdkObjectMap.put(MOLECULER_WEIGHT, new WeightDescriptor());
+        try {
+            // MOLECULER_WEIGHT has 1 component
```
you missed the typo here

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -54,44 +101,533 @@ public enum Descriptor {
+            descriptorToComponentNumberMap.put(FRACTIONAL_PSA, 1);
+            descriptorToCdkObjectMap.put(FRACTIONAL_PSA, new FractionalPSADescriptor());
+
+
```
Unnecessary empty line

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -54,44 +101,533 @@ public enum Descriptor {
+        } catch (Exception anException) {
+            Descriptor.LOGGER.log(
+                    Level.SEVERE,
+                    "Failed to initialize descriptors", anException
```
TODO: we need to discuss what to do in such a case. Right now, the fail behaviour is not optimal. The static initialiser will fail "quietly", the exception is logged but nothing else happens until someone tries to calculate descriptors.

JonasSchaub replied:
Still valid, please put this on the agenda for our next meeting with Achim.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -820,14 +1477,186 @@ private static boolean setDescriptorNew(
     ) {
```
TODO: discuss with Achim what else to call this method.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -820,14 +1477,186 @@ private static boolean setDescriptorNew(
+                    singleBondDesc.setParameters(new Object[]{"s"});
+                    aVector[aStartIndex] = (float) ((IntegerResult) singleBondDesc.calculate(anAtomContainer).getValue()).intValue();
+
+                    BondCountDescriptor doubleBondDesc = new BondCountDescriptor();
```
Can't you re-use the same object and only override the parameter?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
 
-    // Add new descriptor information here!
+    /**
+     * Sets atom count
```
Another JavaDoc formatting thing: the first sentence (must end with .) is taken as the the short description of the method/class. So, you should add a dot here.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+     * Note: This method requires that the implicit hydrogen counts on atoms are properly set
+     *  before calling this method. If implicit hydrogen counts are not set or are all zero,
+     *  no explicit hydrogens will be added to the resulting molecule. Checks are NOT performed here.
+     *  All necessary checks have already been made in public methods above.
```
This method here is also public, so it should perform parameter checks

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+     */
+    public static IAtomContainer createMoleculeWithExplicitHydrogens(
+            IAtomContainer aMolecule
+    ) throws Exception {
```
Please throw a more specific exception than just Exception here.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+            // Copy properties
+            for (Object key : aMolecule.getProperties().keySet()) {
+                tmpMoleculeWithExplicitH.setProperty(key, aMolecule.getProperty(key));
+            }
```
If you copy properties like that, it is not a deep copy anymore because the property values can be references to objects. Do you need to copy specific properties?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+
+                // Copy atom properties
+                for (Object key : tmpAtom.getProperties().keySet()) {
+                    tmpNewAtom.setProperty(key, tmpAtom.getProperty(key));
```
See above

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+                // Copy bond properties
+                for (Object key : tmpBond.getProperties().keySet()) {
+                    tmpNewBond.setProperty(key, tmpBond.getProperty(key));
+                }
```
see above

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+
+                // Add bond to new container
+                tmpMoleculeWithExplicitH.addBond(tmpNewBond);
+            }
```
as discussed, please separate the copying into a specific method, so that you can reuse it.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+
+            // Now that we have a copy of the original molecule, add explicit hydrogens
+            try {
+                AtomContainerManipulator.convertImplicitToExplicitHydrogens(tmpMoleculeWithExplicitH);
```
The method has no throws annotation, did you experience problems here?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+                        "Descriptor.createMoleculeWithExplicitHydrogens: Error converting implicit to explicit hydrogens: " + anException.getMessage(),
+                        anException
+                );
+                throw new Exception("Failed to convert implicit to explicit hydrogens: " + anException.getMessage());
```
In general, catching and re-throwing an exception is considered bad style.

JonasSchaub replied:
And is it also caught and re-thrown again below?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+    /**
+     * Uses a specified aromaticity model to modify aMolecule.
+     * Note: This method changes the input molecule by applying the specified aromaticity model.
+     * Note: No checks are performed, all necessary checks have already been made in public methods above.
```
see above, also for cases in methods below

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+     * @param anAromaticityModel The aromaticity model that will be used for aromaticity detection
+     * @throws Exception Thrown if the aromaticity detection fails or if the aromaticity model is null
+     */
+    public static void setAromaticity(IAtomContainer aMolecule, ElectronDonation anAromaticityModel) throws Exception {
```
see above, also for cases in methods below.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+                        Level.SEVERE,
+                        "Descriptor.setAromaticity: Aromaticity model cannot be null."
+                );
+                throw new IllegalArgumentException("Aromaticity model cannot be null.");
```
adding to the comment above: in general, you either log an exception and then go on or you throw an exception and let the calling code handle/log it.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1924,926 @@ private static synchronized void setWienerNumber(
+            }
+            for (IBond bond : aMolecule.bonds()) {
+                bond.setIsAromatic(false);
+            }
```
The Aromaticity class has a clear() method for this.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -55,35 +61,3327 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
+    }
+
+    /**
+     * Tests method for descriptor H_BOND_ACCEPTOR_COUNT_1
```
There is only one h bond acceptor count descriptor, isn't there? Why _1 and _2 below? Or did you mean to only number the test mehtods?

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -108,83 +3413,452 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
+                                tmpIsParallelCalculation
+                        )
+                );
+                Assertions.assertEquals(8, tmpMatrix[0][0]);
```
It is the same value for all models (8)? What would be the result if not aromaticity detection was performed beforehand? Also, pi system != aromatic system, so I would expect asromaticity to not play a role here.

JonasSchaub reviewed (commented):
Puh, that was a lot of work you did there, well done! I only have minor comments and a few things to discuss. Sorry for sending this review on Friday evening, it can wait until Monday.

ManuelSchauer committed:
Better comments. Separated the copying of a Molecule in createMoleculeWithExplicitHydrogens into the new method copyMolecule (is tested). Small changes to setAromaticity.

ManuelSchauer committed:
Added BASIC_GROUP_COUNT and ACIDIC_GROUP_COUNT descriptors with tests. Slight changes to the test methods.

ManuelSchauer committed:
Added AMINO_ACID_COUNT, added dependency, small changes.

ManuelSchauer committed:
Added explicit hydrogens to HYBRIDIZATION_RATIO, cleaned-up tests.

ManuelSchauer committed:
Added all aromaticity models to the tests for the descriptors that need aromaticity to be set beforehand. Changed default parameters on LARGEST_PI_SYSTEM and BCUT so that checkAromaticity = false.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +1997,986 @@ private static synchronized void setWienerNumber(
+            aVector[aStartIndex + 2] = (float) ((IntegerResult) bondCountDesc.calculate(anAtomContainer).getValue()).intValue();
+        } catch (Exception e) {
+            for (int i = 0; i < 3; i++) {
+                aVector[aStartIndex + i] = Float.NaN;
```
idea to discuss: all these methods return void right now. Maybe we could change that so that they instead return the matrix positions that they marked as NaN if calculation failed.

ManuelSchauer committed:
Added KIER_HALL_SMARTS Descriptor (tested), changed SMALL_RING to include all specific ring sizes.

ManuelSchauer committed:
Added LinkedList as a parameter that tracks NaN values as [moleculeIndex, componentIndex] pairs.

JonasSchaub commented:
@ManuelSchauer please have a look at the medium severity issues SonarCloud reported: https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&pullRequest=4&types=BUG&severities=BLOCKER%2CCRITICAL%2CMAJOR%2CMINOR&inNewCodePeriod=true&sinceLeakPeriod=true&id=JonasSchaub_CDK-Descriptor-Calculation

ManuelSchauer committed:
Change the aNanPositions parameter to be able to accept Collections.synchronizedList() as a parameter if anIsParallelCalculation = true to avoid race conditions. Updated documentation.

ManuelSchauer committed:
Added ECCENTRIC_CONNECTIVITY_INDEX (tested)

ManuelSchauer committed:
Small changes, added two more Descriptors (VABC, MDE).

ManuelSchauer committed:
Changed aNaNPositions List implementation so that no null is allowed.
Minor changes to the logging massages so that they match the given method.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -54,44 +110,635 @@ public enum Descriptor {
+     * 9. C4SP3 - singly bound carbon bound to four other carbons
+     */
+    CARBON_TYPES,
+    //<editor-fold desc="LogP descriptors">
```
Mannhold LogP is not included in the code fold.

ManuelSchauer replied:
Changed.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -54,44 +110,635 @@ public enum Descriptor {
+    /**
+     * APol descriptor, calculates the sum of the atomic polarizabilities (including implicit hydrogens).
+     */
+    A_POL,
```
Maybe polarizability descriptors would be another useful subgroup?

JonasSchaub committed:
mostly minor documentation revisions

JonasSchaub committed:
more doc improvements

JonasSchaub committed:
add exception logging

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -920,7 +2257,1042 @@ private static synchronized void setWienerNumber(
+
+    // Add new descriptor information here!
+    //</editor-fold>
+    //<editor-fold desc="Public static molecule preparation methods">
```
Do they need to be public? If so, please move them on top to the other public methods.

JonasSchaub committed:
finished changes in Descriptor class

JonasSchaub reviewed (commented):
Review of Descriptor class. Review for test class will follow.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -55,214 +63,4106 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
+                            aNanPositions
+                    )
+            );
+            Assertions.assertEquals("8", tmpFormat.format(tmpMatrix[0][0])); // Acetic acid has 8 atoms
```
Is it the expected behaviour that implicit hydrogen atoms are also counted?

JonasSchaub replied:
This should be noted in the documentation of this descriptor enum constant.

ManuelSchauer replied:
Yes it is expected. Changed the documentation

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -55,214 +63,4106 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
+        IAtomContainer tmpMolecule = tmpSmilesParser.parseSmiles(tmpSmiles);
+        IAtomContainer[] tmpMoleculesArray = new IAtomContainer[]{tmpMolecule};
+        int tmpStartIndex = 0;
+        Descriptor[] tmpDescriptors = new Descriptor[]{Descriptor.ATOM_COUNT};
```
I see now that you can configure this descriptor to count specific elements. So, we could introduce another descriptor that uses it to count all atoms of the organic element subset individually (like the bond counts descriptor that counts each bond type separately).

ManuelSchauer replied:
Added.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -55,214 +63,4106 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
+    public void test_H_BOND_ACCEPTOR_COUNT() throws Exception {
+        // Acetic acid
+        String tmpSmiles1 = "CC(=O)O";
+        String tmpSmiles2 = "O=N(=O)c1cccc2cn[nH]c12";
```
name this molecule, please

JonasSchaub replied:
Please do it for all your test molecules

JonasSchaub replied:
instead of the name, you can also give PubChem CID or COCONUT CNP identifiers as "source"

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -55,214 +63,4106 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
+     */
+    @Test
+    public void test_AUTOCORRELATION_CHARGE() throws Exception {
+        String tmpSmiles = "Clc1ccccc1";
```
Maybe use a different test molecule that does not have zero values everywhere?

ManuelSchauer replied:
Discussed in our meeting.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -55,214 +63,4106 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
+     */
+    @Test
+    public void test_AUTOCORRELATION_MASS() throws Exception {
+        String tmpSmiles = "Clc1ccccc1";
```
same here

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -55,214 +63,4106 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
+                    )
+            );
+            //System.out.println("Synchronized calculation results:");
+            //System.out.println(Arrays.toString(tmpMatrix[0]));
```
Maybe at least test whether the results are the same for every parallelization method?

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -55,214 +63,4106 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
+     */
+    @Test
+    public void test_AMINO_ACID_COUNT() throws Exception {
+        String tmpSmiles = "N[C@@]([H])([C@]([H])(O)C)C(=O)N[C@@]([H])([C@]([H])(O)C)C(=O)O"; // protein TT
```
I'd like to know whether stereochemistry and explicit hydrogens *have to be* provided in order for it to work.

ManuelSchauer replied:
It works without, added Glyglycin without it and it works as intended.

JonasSchaub reviewed (commented):
Review of test class

JonasSchaub committed:
mostly minor doc improvements

ManuelSchauer committed:
Merge pull request #6 from JonasSchaub/review-dev-manuel

ManuelSchauer committed:
Added ATOM_COUN_ORAGNIC_SUBSET

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -641,6 +647,10 @@ public enum Descriptor {
 
+            // ATOM_COUNT_ORGANIC_SUBSET has 1 component
+            descriptorToComponentNumberMap.put(ATOM_COUNT_ORGANIC_SUBSET, 10);
+            descriptorToCdkObjectMap.put(ATOM_COUNT_ORGANIC_SUBSET, new AtomCountDescriptor());
```
I think you missed a "0" in the comment :D

ManuelSchauer committed:
Added a new molecule to test_AMINO_ACID_COUNT()

ManuelSchauer committed:
Added isFast and isSafe properties to Descriptor enum

- Added constructor with boolean parameters for speed and safety characteristics
- Added isFast() method getter for isFast property
- Added isSafe() method getter for isSafe property
- Update all enum constants with appropriate fast/safe values
- Added documentation

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -871,24 +920,43 @@ public static Descriptor[] getAllDescriptors(){
-            return null;
         }
+
+        java.util.List<Descriptor> result = new java.util.ArrayList<>();
```
ArrayList should be initialised with a proper initial size (because the underlying array needs to be re-computed for every re-sizing and that takes time)!

ManuelSchauer replied:
Changed it.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -3481,6 +3549,7 @@ private static synchronized void setVABC(
      * Creates a deep copy of the input molecule.
      * Note: This method is used to create a new molecule object
      * without affecting implicit hydrogen atoms.
+     * Note: If necessary Atom types must be perceived and configured manually after creation.
```
Put a comma after "If necessary" and start atom lower-case

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -871,24 +920,43 @@ public static Descriptor[] getAllDescriptors(){
      * @param isQuicklyCalculableDescriptorInclusion True: Quickly calculable descriptors are returned, false: Otherwise.
      * @param isSlowlyCalculableDescriptorInclusion True: Slowly calculable descriptors are returned, false: Otherwise.
+     * @param excludeUnsafeDescriptors True: Unsafe descriptors are excluded from the result, this does not mean no
+     *                                 NaN's can be produced, false: Otherwise.
```
Sorry but I would like you to turn this around, to be consistent with the other parameters, i.e. define it as "isUnsafeDescriptorInclusion".

ManuelSchauer replied:
Changed it.

ManuelSchauer committed:
Changed excludeUnsafeDescriptors to isUnsafeDescriptorInclusion and isFast of AMINO_ACID_COUNT to false.

ManuelSchauer committed:
Changed ArrayList initialization to a constant value of the maximum possible capacity.

ManuelSchauer committed:
Array initialization changed in test_Parallelization() and test_Integrity().
Minor comment enhancements.

ManuelSchauer committed:
Changed descriptorToComponentNumberMap to an enum property descriptorComponentNumber.

ManuelSchauer committed:
Updated SMILES strings in DescriptorTest with corresponding PubChem CIDs and names if possible.

ManuelSchauer committed:
Added DescriptorCalculator for non-synchronized parallelization calculation.
- Added disabled test for DescriptorCalculator

Changed naming convention in synchronized method.

ManuelSchauer committed:
Addend an integrity test for DescriptorCalculator.
Added another test to verify problematic descriptors.

ManuelSchauer committed:
Changed ATOM_COUNT_ORGANIC_SUBSET and BOND_COUNT_SPECIFIED to have a separate objects for each parameter to avoid race conditions

ManuelSchauer committed:
- Added fingerprints as descriptors.
- Descriptors with parameters in ATOM_COUNT_ORGANIC_SUBSET and BOND_COUNT_SPECIFIED are now implemented individually

ManuelSchauer committed:
"Unsafe" method now included in the main class, updated all tests with the new method

ManuelSchauer committed:
Updated to CDK-Snapshot,
added tests for fingerprinter and minor doc changes

ManuelSchauer committed:
Merge branch 'production' into dev-manuel

ManuelSchauer committed:
Implemented Batch-Processing, updated parallelization and integrity tests

ManuelSchauer committed:
Merge remote-tracking branch 'origin/dev-manuel' into dev-manuel

ManuelSchauer committed:
Circular fingerprinter implementation updated: ECFP/FCFP variants for all path diameters (0, 2, 4, 6) now implemented individually

ManuelSchauer committed:
Added batch calculation with a string array, minor doc improvements

ManuelSchauer committed:
minor doc improvements

JonasSchaub commented:
@ManuelSchauer you should add your name in the Licence-header.txt file.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
 import java.util.stream.IntStream;
 
 /**
  * Descriptor related calculations based on the CDK for the enrichment of data vectors.
```
Descriptor-related

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
  * Descriptor related calculations based on the CDK for the enrichment of data vectors.
- * Note. There are 3 "public static boolean setDescriptorsForMolecules...()" methods with different forms of
+ * Note: There are 3 "public static boolean setDescriptorsForMolecules...()" methods with different forms of
  * (parallelized) calculation since individual CDK descriptor calculation classes are unfortunately NOT thread-safe.
```
Isn't it 4 now? And the CDK descriptor classes are thread-safe now!

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
+ * IAtomContainer molecule = ...;
+ *
+ * // 2. IMPORTANT: Perform aromaticity perception before descriptor calculation
+ * Descriptor.setAromaticity(molecule);
```
This example is incomplete, an aromaticity model has to be specified.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
+ *     Descriptor.A_LOG_P
+ * };
+ * // or use a method to get specific descriptors:
+ * Descriptor[] descriptors = Descriptor.getSpecifiedDescriptors(); // e.g., user-defined selection
```
also here, add parameters in the form of boolean variables with suitable names that you define before calling the method.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
+    //<editor-fold desc="Basic Bond and Count descriptors">
     /**
-     * EnumMap that maps a descriptor to its number of calculated components
+     * Atom count, counts the number of all atoms in the given molecule.
```
All heavy atoms or also hydrogen atoms? Please specify.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
+     *
+     * @see AtomCountDescriptor
+     */
+    ATOM_COUNT_H(true, true, 1),
```
only explicit or also implicit H atoms? Please specify.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
+    /**
+     * Total bond count, counts the number of all bonds in a molecule, neglecting the order.
+     * Double and triple bonds are counted as one bond.
+     * Bonds to hydrogen atoms are counted.
```
Only to explicit H atoms or also implicit ones? please specify.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
+     * @return All available descriptors
+     */
+    public static Descriptor[] getAllDescriptors(){
+        return values();
```
should this maybe excluded fingerprints? To discuss.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
+
+    /**
+     * Returns specified available descriptors.
+     * Note: If both speed flags are false, an empty array is returned. If both speed flags are true, then all descriptors are returned.
```
Update the doc here. E.g. the third flag makes it more complex than it is stated here.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
+            return new Descriptor[0];
+        }
+        // Initialize ArrayList with maximum possible capacity to avoid internal resizing during element addition
+        java.util.List<Descriptor> result = new java.util.ArrayList<>(values().length);
```
omit package paths

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -25,120 +25,1210 @@
+    }
+
+    /**
+     * Returns sum of number of calculated components of an array of defined descriptors.
```
This could be more descriptive, e.g. returns the total number of required components (i.e. matrix columns) if all given descriptors are calculated for a given molecule or molecule set.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -163,304 +1253,624 @@ public static int getNumberOfComponents(
+                "Descriptor.getNumberOfComponents: Number of components could not be evaluated: This should never happen.", anException
             );
-            throw new Exception("Descriptor.getNumberOfComponents: An exception occurred: This should never happen.");
+            throw new Exception("Descriptor.getNumberOfComponents: An exception occurred: This should never happen.", anException);
```
This was probably Achim und not you but the general rule is to either log an exception or (re-)throw it. That is because the calling code above will also handle the exception (i.e. log it in most cases), so you don't need to do it here.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -163,304 +1253,624 @@ public static int getNumberOfComponents(
+        }
+        //</editor-fold>
+        try {
+            AtomContainerManipulator.percieveAtomTypesAndConfigureAtoms(aMolecule);
```
This is only necessary if the given electron donation model is CDK_AtomTypes.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -163,304 +1253,624 @@ public static int getNumberOfComponents(
+            Cycles.markRingAtomsAndBonds(aMolecule);
+            Aromaticity.apply(anAromaticityModel, aMolecule);
+            CDKHydrogenAdder adder = CDKHydrogenAdder.getInstance(aMolecule.getBuilder());
+            adder.addImplicitHydrogens(aMolecule);
```
Why are you adding implicit hydrogen atoms here?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -163,304 +1253,624 @@ public static int getNumberOfComponents(
+            CDKHydrogenAdder adder = CDKHydrogenAdder.getInstance(aMolecule.getBuilder());
+            adder.addImplicitHydrogens(aMolecule);
+        } catch (Exception anException) {
+            throw new Exception("Failed to detect aromaticity: " + anException.getMessage(), anException);
```
You are not really adding any information to the exception and simply re-throwing it. For that, you need no catch block.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+     * @param descriptor The descriptor to check
+     * @return true if the descriptor is a fingerprint, false otherwise
+     */
+    static boolean isFingerprint(Descriptor descriptor) {
```
To discuss whether this should be another enum property instead of this list.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -163,304 +1253,624 @@ public static int getNumberOfComponents(
+    /**
+     * Sets calculated descriptor components in vectors (rows) of a aMatrix (that corresponds to anAtomContainerArray)
+     * beginning with aStartIndex by (optional) parallelization of molecule batches.
+     * Note: Fingerprint descriptors are not supported.
```
add to the doc that descriptor class instances are accessed by multiple threads in parallel.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -163,304 +1253,624 @@ public static int getNumberOfComponents(
-                        tmpIsSuccessful = false;
+                for (int i = 0; i < tmpNumberOfMolecules; i++) {
+                    if (!Descriptor.setDescriptorsForSingleMolecule(aDescriptors, anAtomContainerArray[i], aMatrix[i], tmpStartIndices, i, aNanPositions)) {
+                        tmpHasNaN.set(true);
```
To discuss: implement one method for serial computation instead of having these four lines and the extra parameter if all the different parallelization methods. Or is there a difference I don't see?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -163,304 +1253,624 @@ public static int getNumberOfComponents(
-     * Sets calculated descriptor components in vectors of a aMatrix (that corresponds to anAtomContainerArray)
-     * beginning with aStartIndex by (optional) parallelization of molecules.
-     * Note: Uses a new descriptor instance for EVERY descriptor calculation which slows down the calculation.
+     * Sets calculated descriptor components in vectors (rows) of a aMatrix (that corresponds to anAtomContainerArray)
```
anAtomContainerArray -> aMoleculeSmilesStringArray

JonasSchaub replied:
and see the other comments on the similar method above.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -163,304 +1253,624 @@ public static int getNumberOfComponents(
      * @param aDescriptors Array of descriptors to be calculated (IS NOT CHANGED)
-     * @param anAtomContainerArray Array of molecules. Note: anAtomContainerArray[i] corresponds to aMatrix[i] data
-     *                              vector. (IS NOT CHANGED)
+     * @param aMoleculeSmilesStringArray Array of molecule SMILES stings. Note: anAtomContainerArray[i] corresponds to aMatrix[i] data
```
stings -> stRings

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -163,304 +1253,624 @@ public static int getNumberOfComponents(
+            }
+            if (anIsParallelCalculation) {
+                try {
+                    boolean[] tmpIsDescriptorCalculations = new boolean[anAtomContainerArray.length];
```
why the array here and not an atomic boolean like in the batch methods?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
-     * or synchronized descriptor calculation are avoided.
+     * Sets calculated descriptor components in vectors (rows) of a aMatrix (that corresponds to anAtomContainerArray)
+     * beginning with aStartIndex by (optional) parallelization of molecules. If parallel computation is used, the atom
+     * containers (molecules) are distributed onto parallel thread, one for each molecule, and they all access shared
```
threadS

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+            }
+            if (anIsParallelCalculation) {
+                try {
+                    boolean[] tmpIsDescriptorCalculations = new boolean[anAtomContainerArray.length];
```
see above.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+    /**
+     * Sets calculated descriptor components in vectors (rows) of a aMatrix (that corresponds to anAtomContainerArray)
+     * beginning with aStartIndex by (optional) parallelization of molecules. If parallel computation is used, the atom
+     * containers (molecules) are distributed onto parallel thread, one for each molecule, and they all access shared
```
threadS

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+     * beginning with aStartIndex by (optional) parallelization of molecules. If parallel computation is used, the atom
+     * containers (molecules) are distributed onto parallel thread, one for each molecule, and they all access shared
+     * descriptor instances.
+     * Note: Uses a new descriptor instance for EVERY descriptor calculation which slows down the calculation.
```
mention the higher memory consumption which is the main point here.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+            }
+            if (anIsParallelCalculation) {
+                try {
+                    boolean[] tmpIsDescriptorCalculations = new boolean[anAtomContainerArray.length];
```
see above.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+    /**
+     * Sets calculated descriptor components in vectors (rows) of a aMatrix (that corresponds to anAtomContainerArray)
+     * beginning with aStartIndex by (optional) parallelization of descriptor calculations. If parallel computation is
+     * used, the descriptor instances are distributed onto parallel thread, one for each descriptor, and they all access shared
```
threadS

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+            boolean anIsParallelCalculation,
+            List<int[]> aNanPositions
+    ) throws IllegalArgumentException, Exception {
+        //<editor-fold desc="Checks">
```
idea: since all these checks are the same for all the calculation methods (I presume), could we implement a central validation method?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+            }
+            if (anIsParallelCalculation) {
+                try {
+                    boolean[] tmpIsMoleculeCalculations = new boolean[aDescriptors.length];
```
see above

JonasSchaub reviewed (commented):
First part of review, reviewing the public methods.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+    ) throws Exception {
+        try {
+            SmilesParser tmpSmilesParser = new SmilesParser(SilentChemObjectBuilder.getInstance());
+            IAtomContainer tmpMolecule = tmpSmilesParser.parseSmiles(aMoleculeSmilesString);
```
Can you try out whether it works with a static SmilesParser instance?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+        try {
+            switch (aDescriptor) {
+                case MOLECULAR_WEIGHT:
+                    aVector[aStartIndex] = (float) ((DoubleResult) descriptorToCdkObjectMap.get(MOLECULAR_WEIGHT).calculate(anAtomContainer).getValue()).doubleValue();
```
Why can't you use the specific descriptor set methods here?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+     * @param bitFingerprint The bit fingerprint to convert
+     * @return Float array with 1.0f for set bits and 0.0f for unset bits
+     */
+    private static float[] convertBitFingerprintToFloatArray(IBitFingerprint bitFingerprint) {
```
It would be more memory-efficient to not allocate another float array for the fingerprint first but instead work directly on the matrix, like in the other methods.

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+     * @throws IllegalArgumentException If the input molecule is empty
+     * @throws CloneNotSupportedException If the molecule cannot be properly copied
+     */
+    static IAtomContainer copyMolecule(IAtomContainer aMolecule)
```
Why are this method and others package-private?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+                tmpNewAtom.setMassNumber(tmpAtom.getMassNumber());
+                tmpNewAtom.setFormalCharge(tmpAtom.getFormalCharge());
+                tmpNewAtom.setImplicitHydrogenCount(tmpAtom.getImplicitHydrogenCount());
+                tmpNewAtom.setCharge(tmpAtom.getCharge());
```
setting the formal charge also sets the (partial) charge, see https://github.com/cdk/cdk/pull/1151

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+                // Add bond to new container
+                tmpMoleculeCopy.addBond(tmpNewBond);
+            }
+            return tmpMoleculeCopy;
```
Copying lone pairs, single electrons, and stereo elements could also be relevant for some descriptors, please check (for code example see https://github.com/cdk/cdk/blob/9c3f48f4b36cd7cde507428ea80d07362c759c25/misc/extra/src/main/java/org/openscience/cdk/tools/SugarDetectionUtility.java#L1020).

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
+     * @param aVector Vector of molecule to be filled with calculated components of descriptors (MAY BE CHANGED)
+     * @param aStartIndex Start index in aVector to be filled with calculated components of descriptors
+     */
+    private static synchronized void setMolecularWeightSynchronized(
```
A more optimal synchronisation that does not block the entire class could be:
```
synchronized (descriptorToCdkObjectMap.get(MOLECULAR_WEIGHT)) {
    aVector[aStartIndex] = (float) ((DoubleResult) descriptorToCdkObjectMap.get(MOLECULAR_WEIGHT).calculate(anAtomContainer).getValue()).doubleValue();
}
```
within the method instead of declaring the method synchronized. If I understand this tutorial correctly: https://www.baeldung.com/java-synchronized

Could you check that?

JonasSchaub commented on src/main/java/de/unijena/cheminf/clustering/desccalc/Descriptor.java:
```diff
@@ -483,444 +1898,4910 @@ public static boolean setDescriptorsForMoleculesByMoleculeParallelizationNew(
-            int[] tmpStartIndices = new int[aDescriptors.length];
-            for (int i = 0; i < aDescriptors.length; i++) {
-                tmpStartIndices[i] = aStartIndex;
-         
```
"This class assumes that aromaticity perception, atom typing and adding of implicit hydrogens have been performed prior to generating the fingerprint." - all that has been done prior to this step, right?

JonasSchaub reviewed (commented):
Final part of review of Descriptor class. 
To discuss: potentially splitting up this massive class.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -42,22 +52,22 @@
+     * Test method for descriptor MOLECULAR_WEIGHT.
      */
     @Test
     public void test_MOLECULAR_WEIGHT() throws Exception {
```
Please remove all public modifiers. JUnit5 test classes and methods should generally have package visibility.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -185,101 +242,5430 @@ public void test_WIENER_NUMBER() throws Exception {
+                    )
+            );
+            //System.out.println("Synchronized calculation results:");
+            //System.out.println(Arrays.toString(tmpMatrix[0]));
```
Please remove all commented-out code and print statements.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -185,101 +242,5430 @@ public void test_WIENER_NUMBER() throws Exception {
+        Descriptor[] tmpDescriptors = new Descriptor[]{Descriptor.PUBCHEM_FINGERPRINTER};
+        boolean tmpIsParallelCalculation = false;
+        // Verify that expected bits are set (convert reference fingerprint to expected values)
+        IFingerprinter printer = new PubchemFingerprinter(SilentChemObjectBuilder.getInstance());
```
unused variable

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -185,101 +242,5430 @@ public void test_WIENER_NUMBER() throws Exception {
+            }
+
+        } catch (Exception anException) {
+            Assertions.fail();
```
This is unnecessary, if there is an exception thrown, the test fails automatically.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -66,46 +76,69 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
+                            aNanPositions
+                    )
             );
             Assertions.assertEquals("60.05", tmpFormat.format(tmpMatrix[0][0]));
```
Normally, I would ask you to put the 60.05 in a variable, since you are re-using it so many times below. But well, just leave it here now...

JonasSchaub replied:
Same for all the other hard-coded results in the other methods...

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -134,49 +167,73 @@ public void test_WIENER_NUMBER() throws Exception {
+                    )
+            );
+            Assertions.assertEquals("9", tmpFormat.format(tmpMatrix[0][0])); // 1(C1C2)+2(C1O1)+2(C1O2)+1(C2O1)+1(C2O2)+2(O1O2) = 9
+            Assertions.assertEquals("0", tmpFormat.format(tmpMatrix[0][1])); // there are no atoms that are 3 bonds apart
```
These are integers that should be compared as such. Same is true for other test methods.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -185,101 +242,5430 @@ public void test_WIENER_NUMBER() throws Exception {
-            Assertions.assertEquals("60.05", tmpFormat.format(tmpMatrix[0][0]));
-            Assertions.assertEquals("9.00", tmpFormat.format(tmpMatrix[0][1]));
-            Assertions.assertEquals("0.00", tmpFormat.format(tmpMatrix[0][2]));
+            Assertions.assertEquals("8", tmpFormat.format(tmpMatrix[0][0])); // Acetic acid has 8 atoms
```
Is there a way to count only heavy atoms? That would actually be my expectation.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -115,17 +148,17 @@ public void test_MOLECULAR_WEIGHT() throws Exception {
 
     /**
-     * Tests method for descriptor WIENER_NUMBER
+     * Tests method for descriptor WIENER_NUMBER.
```
Test, not Tests. Same in other test methods.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -185,101 +242,5430 @@ public void test_WIENER_NUMBER() throws Exception {
+            Assertions.assertEquals(-0.381844, tmpMatrix[0][2], epsilon);
+            Assertions.assertEquals(0.325509, tmpMatrix[0][3], epsilon);
+            Assertions.assertEquals(3.374638, tmpMatrix[0][4], epsilon);
+            Assertions.assertEquals(5.033583, tmpMatrix[0][5], epsilon);
```
Here, you are testing it in the correct way! Would it be too much to ask to do it in the other places this way as well?

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -185,101 +242,5430 @@ public void test_WIENER_NUMBER() throws Exception {
+            Assertions.assertNotEquals(0f, tmpMatrix[0][1]);
+            Assertions.assertNotEquals(0f, tmpMatrix[0][2]);
+            Assertions.assertNotEquals(0f, tmpMatrix[0][3]);
+            Assertions.assertNotEquals(0f, tmpMatrix[0][4]);
```
Maybe use a molecule whose expected avlues are not all 0?

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -185,101 +242,5430 @@ public void test_WIENER_NUMBER() throws Exception {
+                            aNanPositions
+                    )
+            );
+            Assertions.assertNotNull(tmpMatrix);
```
This check doesn't make much sense. Maybe rather test that the NaN value list is empty?

JonasSchaub replied:
Same below, in the other tests

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -471,4 +6142,184 @@ public void test_Integrity() throws Exception {
+
+            // Set aromaticity using the current model
+            // Note: This test still uses the old Descriptor.setAromaticity which takes ElectronDonation
+            // It might need updating if Descriptor.setAromaticity now expects Aromaticity.Model
```
No need for that note, the Aromaticity.Model instances are also of class ElectronDonation.

JonasSchaub commented on src/test/java/de/unijena/cheminf/clustering/desccalc/DescriptorTest.java:
```diff
@@ -471,4 +6142,184 @@ public void test_Integrity() throws Exception {
+     * Tests setAromaticity.
+     */
+    @Test
+    public void testSetAromaticity() throws Exception {
```
I think this test is out of scope for this class. You're simply testing whether all CDK aromaticity models detect benzene as aromatic.

JonasSchaub reviewed (commented):
Review of test class.

JonasSchaub commented:
@ManuelSchauer can we merge this PR and you implement the requested changes afterwards (you can use the same branch)? My browser is always close to crashing whenever I load the changes...

ManuelSchauer committed:
removed public modifier in test class and changed implementation of the smiles parser

ManuelSchauer committed:
added isFingerprint as enum constant, added abstract calculation method -> changed calculation logic in setDescriptorMap, setDescriptorSynchronized and setDescriptor

ManuelSchauer committed:
updated test class to number comparisons and LICENSE.txt

ManuelSchauer committed:
implemented new enum constant needsExplicitHydrogens which is now handled in the setDescriptorsForSingleMolecule methods

ManuelSchauer committed:
updated circular fingerprinter tests, small changes
